### DISCUSSION TITLE
Prisma & MongoDb

### DISCUSSION THREAD

**Discussion by CodeRider101** [SE-TINF22B2/G2-Schoolify#59] [Technologien]
Warum benutzen wir Prisma & MongoDb und nicht andere Technologien auf dieser Ebene?
@Kronprinz03

**Comment by Kronprinz03** [comment]
### Erklärung Prisma:
Prisma ist ein Open-Source-ORM, das es ermöglicht, die Datenbank im Code zu definieren (Beziehungen, Attribute und Entitäten). Prisma verbessert die Lesbarkeit und unterstützt die Nutzung von TypeScript, da die Datenbank im Code definiert wird, und bietet ein typsicheres Query-Builder-System. Änderungen an den Datenstrukturen können auch direkt auf die Datenbank übertragen werden.

### Erklärung MongoDB:
Das Dokumentmodell von MongoDB Atlas ermöglicht Entwicklern, Daten in Form von JSON-ähnlichen Objekten zu speichern, die Objekten im Anwendungscode ähneln. Hierfür stehen der MongoClient und Mongoose zur Verfügung.

### Kompatibilität:

Nest.js -> [x] Prisma [x] MongoDB
MongoDB -> [x] Prisma [x] MongoDB
Prisma -> [x] Prisma [x] MongoDB

### Beurteilung:

Nest.js + Prisma:
Diese Kombination von Technologien ist sinnvoll, da Nest.js mit seinen Validierungspipelines und TypeScript die Typsicherheit auch auf die Datenbankebene erweitert. Prisma vereinfacht die Programmierung, da es Entitäten direkt im Code vorschlägt, was die Fehleranfälligkeit verringert. Als Alternative könnte man Mongoose verwenden, jedoch nur, wenn MongoDB Atlas verwendet wird.

MongoDB + Prisma:
Diese Verbindung ist nicht besonders sinnvoll, da Prisma relationale Datenbanken definiert, während MongoDB Atlas für dokumentenorientierte Datenbanken ausgelegt ist. Daher passen MongoDB und Prisma nicht gut zusammen und sollten nicht unbedingt gemeinsam genutzt werden. Hier wäre es sinnvoller, das Node-Paket Mongoose zu verwenden.

MongoDB Atlas:
Es ist sinnvoll, für dieses Projekt eine nicht-relationale Datenbank zu nutzen. Es bringt keinen Vorteil, eine relationale Datenbank zu verwenden, da die Beziehungen eindeutig sind. Die nicht-relationale Datenbank hat keine offensichtlichen Nachteile, außer dass die Verbindung mit Prisma nicht sinnvoll ist.

### Fazit:
Man muss sich jetzt entscheiden, ob man Prisma mit einer relationalen Datenbank wie MySQL oder PostgreSQL oder MongoDB mit Mongoose verwenden möchtest. Hinsichtlich der Typsicherheit wäre die Verwendung von Prisma erstrebenswert.
